dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/rest/rest/client/user/MessageAndRestClientTestServlet.java

```diff
@@ -0,0 +1,145 @@
+
+    @Inject
+    private SimpleReactiveMessagingBean reactiveBean;
+
+    @Inject
+    InMemorySpanExporter inMemorySpanExporter;
+
+    @Test
+    public void testReactiveMessagingThatCallsJaxClient() throws InterruptedException {
+        List<String> results = getResultsWithWaitIfNeeded();
+
+        assertEquals(3, results.size());
+        assertEquals(3, results.size());
+        assertFalse(results.contains("LENGTH 8"));
+        assertFalse(results.contains("LENGTH 9!"));
+        assertTrue(results.contains("LENGTH 10!"));
+        assertTrue(results.contains("LENGTH 11!!"));
+        assertTrue(results.contains("LENGTH 12!!!"));
+    }
+
+    @Test
+    public void testSpanInBothJaxCallerAndCalled() throws InterruptedException {
+        getResultsWithWaitIfNeeded();//we just want to delay until the results are in.
+        List<List<SpanData>> listOfListOfResults = new ArrayList<List<SpanData>>(5);
+
+        SpanDataMatcher spanInsideClientMatcher = SpanDataMatcher.hasName("SimpleReactiveMessagingBean.toUpperCase");
+
+        SpanDataMatcher spanInsideJaxServiceMatcher = SpanDataMatcher.hasName("RestClientThatCapitalizes.jaxServiceThatCapitalizes");
+
+        //Each message will end up with four spans being registered. And there are five messages.
+        for (Span s : spanAndExpectedNumber.keySet()) {
+            listOfListOfResults.add(new ArrayList(inMemorySpanExporter.getFinishedSpanItems(spanAndExpectedNumber.get(s), s)));
+        }
+
+        String failMessage = "";
+
+        //We don't really need to test this five times, but keeping all five
+        //makes the test app more authentic and since we have them.
+        for (int index = 0; index < listOfListOfResults.size(); index++) {
+            boolean foundSpanCallingJax = false;
+            boolean foundSpanInsideJax = false;
+
+            for (SpanData sd : listOfListOfResults.get(index)) {
+                if (spanInsideClientMatcher.matches(sd)) {
+                    foundSpanCallingJax = true;
+                }
+
+                if (spanInsideJaxServiceMatcher.matches(sd)) {
+                    foundSpanInsideJax = true;
+                }
+
+                if (foundSpanCallingJax && foundSpanInsideJax) {
+                    break;
+                }
+            }
+            if (!foundSpanCallingJax || !foundSpanInsideJax) {
+                StringBuilder sb = new StringBuilder();
+                sb.append("failed on index: " + index);
+                if (!foundSpanCallingJax) {
+                    sb.append(" Did not find a Span from SimpleReactiveMessagingBean");
+                }
+                if (!foundSpanInsideJax) {
+                    sb.append(" Did not find a Span from RestClientThatCapitalizes");
+                }
+                sb.append(System.lineSeparator());
+                failMessage = failMessage + sb.toString();
+            }
```

Comment:
Kinda torn on this one. I guess you're ensuring that you're reporting all the results?

It's annoying that it's so much more code than just doing a basic check like this:

```java
assertThat("Failed to find spans for index " + index,
      listOfListOfResults.get(index),
      allOf(hasItem(spanInsideClientMatcher),
            hasItem(spanInsideJaxServiceMatcher()));
```

Or maybe more readably like this:
```java
List<SpanData> results = listOfListOfResults.get(index);
assertThat("Failed to find spans for index "+ index, results, hasItem(span().withName("SimpleReactiveMessagingBean.toUpperCase")));
assertThat("Failed to find spans for index "+ index, results, hasItem(span().withName("RestClientThatCapitalizes.jaxServiceThatCapitalizes")));
```

You could try to wrap up all the checks into one long assert but the error message there would probably be a bit incomprehensible so it's probably best left as is, even though it's not easy to see what the test is checking.